plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/expressions/FirKotlinUTypeReferenceExpression.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.PsiType
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UTypeReferenceExpression
+import org.jetbrains.uast.UastErrorType
+
+class FirKotlinUTypeReferenceExpression(
+    override val sourcePsi: KtTypeReference,
+    givenParent: UElement?,
+    private val typeSupplier: (() -> PsiType)? = null,
+) : FirKotlinAbstractUExpression(givenParent), UTypeReferenceExpression {
+    override val type: PsiType by lz {
+        typeSupplier?.invoke() ?: sourcePsi.toPsiType(uastParent ?: this)
+    }
+}
+
+internal fun ConeKotlinType.toPsiType(): PsiType {
+    // TODO
```

Comment:
FYI, @darthorimar, perhaps my next prototyping work would be utilizing/refactoring the type conversion, `ConeKotlinType#asPsiType(...)` in `firLightUtils.kt`.

Reply:
Should we be using `KtType` over `ConeType` here?

Reply:
You're right; it's not necessary to reveal `ConeType` at the moment. I skimmed a bit, and... perhaps we don't even need to disclose `KtType`?! We'll see.